### ISSUE TITLE
Executable is in the middle of the heap

### ISSUE THREAD

**Issue by HW42** [oscarlab/graphene#583]
Example for the `Bootstrap` regression test:
```
adding page  to enclave: 000000000ffff000 [REG:R--] (manifest) measured
adding pages to enclave: 000000000fff6000-000000000fffe000 [REG:RW-] (ssa)
adding pages to enclave: 000000000fff1000-000000000fff5000 [TCS:---] (tcs)
adding pages to enclave: 000000000ffec000-000000000fff0000 [REG:RW-] (tls)
adding pages to enclave: 000000000ffdb000-000000000ffeb000 [REG:RW-] (stack)
adding pages to enclave: 000000000ffca000-000000000ffda000 [REG:RW-] (stack)
adding pages to enclave: 000000000ffb9000-000000000ffc9000 [REG:RW-] (stack)
adding pages to enclave: 000000000ffa8000-000000000ffb8000 [REG:RW-] (stack)
adding pages to enclave: 000000000bd76000-000000000bda3000 [REG:R-X] (code) measured
adding pages to enclave: 000000000bfa2000-000000000bfa4000 [REG:RW-] (data) measured
adding pages to enclave: 000000000bfa4000-000000000ffa7000 [REG:RW-] (bss)
adding page  to enclave: 0000000000400000 [REG:RWX] (code) measured
adding pages to enclave: 0000000000600000-0000000000602000 [REG:RW-] (data) measured
adding pages to enclave: 0000000000602000-000000000bd75000 [REG:RWX] (free)
adding pages to enclave: 0000000000010000-0000000000400000 [REG:RWX] (free)
```

The heap goes from 0x10000 to 0xbd75000. But there's the executable in the middle of it (0x400000 to 0x602000).

There's code in `init_pages` to handle it, so this works. But what's the point of this design? It makes things like zeroing the heap more tricky to implement. Also there is no memory gap between the executable and the heap (unlike the other memory areas).

**Comment by dimakuv** [comment]
The idea of the code (`sgx_main.c: initialize_enclave()`) is to add the following memory areas to the enclave: manifest, SSA, TCS, TLS, stack for each enclave thread, code + data + bss for PAL library, code + data + bss for actual app. The memory areas are added one by one starting from the top of the enclave (in the example with `bootstrap` it is 0x10000000).

Graphene has a requirement to support non-PIE code (`bootstrap` is non-PIE while `bootstrap_pie` is PIE). So Graphene puts `bootstrap`'s code based on the executable's code segment -- which is 0x400000 by default in gcc. Interestingly, Graphene also puts the data segment based on the executable's data segment -- which is 0x600000 by default. All memory which the executable uses is merged into one `exec_area` -- thus we see a hole between 0x400000 and 0x600000 (*which can be considered a bug since we lose this memory region which could be used for heap*). (This is all reading SGX PAL's `sgx_main.c: load_enclave_binary()`.)

This requirement creates a discrepancy between the "who-cares-where" pages which are put at the top of the enclave and the "put-us-where-told" executable code+data pages.

Now, Graphene needs to add free pages -- to form the heap. Since we want to use all available enclave memory, the logic just takes the two holes: `[exec.base+exec.size, topmost-available-page)` `[heap-min, exec.base)` (heap-min is usually set to 0x10000 by Graphene). These are the two FREE regions we see at the end. They kinda wrap around the unavailable `exec_area` (code+data regions before).

Regarding the missing memory gap: this is just a curiosity of the code:
```c
/* first FREE region */
unsigned long addr = exec_area->addr + exec_area->size;
if (addr < heap_min)  addr = heap_min;
set_area("free", true, false, -1, addr, populating - addr, ...);  // should be ..., addr + MEMORY_GAP, ...
/* second FREE region */
set_area("free", true, false, -1, heap_min, populating - heap_min, ...); // should be ..., populating - heap_min - MEMORY_GAP, ...
```

Finally, as you already mentioned, the logic in `init_pages()` marks the `exec_area` as unavailable for heap memory by adding a dummy VMA which is never freed. This indeed looks like a hack (code segment inside of heap?!), but not sure what else can be done.

So, to conclude this exploration:
1. There is no simple way to have a continuous heap area because of the "put-executable-at-specific-address" requirement (for non-PIE executables). And carving out only one heap area -- either before or after this executable area -- significantly reduces available heap size.
2. The memory gap issue is minor (not a real bug but detrimental to memory safety) and can be easily fixed. See my code above.
3. The zeroing-the-heap code could be implemented as follows: go through `[pal_sec.heap_min, pal_sec.heap_max)` but exclude all VMAs in `heap_vma_list` (to not accidentally zero-out the executable area). Not too elegant but works.

P.S. What the heck happens with BSS segment of PAL? It stretches from 0xbfa4000 to 0xffa7000 which is a region of 64MB. This could be used for the heap!

**Comment by HW42** [comment]
Ok, thanks for looking at this.

So to the answer to my question "Why this design" is: non-PIE code + "we want to use all available enclave memory". So everything working as intended.

The memory gap thing is a bug.

> P.S. What the heck happens with BSS segment of PAL? It stretches from 0xbfa4000 to 0xffa7000 which is a region of 64MB. This could be used for the heap!

Hah, I also noticed this one today. The answer is that the "slab" code has some statically allocated memory pool (see `Pal/src/slab.c`).

Since zeroing the heap is faster than measuring zeros I tried to drop this. Unfortunately `STATIC_SLAB == 0` is broken. I did not look too closely but I think the problem is that `create_slab_mgr` calls `__malloc` which calls `_DkVirtualMemoryAlloc` which calls `get_reserve_pages` which calls `malloc` ... And currently this loop is broken by the static mem pool.

**Comment by HW42** [comment]
Re the memory gap:

Do you know what's the purpose of this code (or it's counterpart in `sgx_main.c`):

```python
        if area.desc == 'exec':
            populating = area.addr;
        else:
            populating = area.addr - MEMORY_GAP
```

This code is only reached if `area.addr` is `None`, which AFAIU is never the case for the exec area. And indeed none of the Pal and shim regression tests touch this line.

**Comment by dimakuv** [comment]
True, this code looks like dead code.

Unless the code segment of the executable explicitly starts at address `0x0`. In this case, `pal-sgx-sign` will have `area.addr = None` and we end up in this code snippet. Similarly, `sgx_main.c` will also end up in this code snippet if `areas[i].addr == 0x0`.

Is this even a thing that executables can start with address `0x0`? I guess no sane binary does that. I would remove this code then and maybe add an assert to fail loudly if `area.desc == 'exec' and area.addr is None` (and similarly in `sgx_main.c`).

P.S. What tools do you use that tell you which lines of tests are touched?

**Comment by HW42** [comment]
> P.S. What tools do you use that tell you which lines of tests are touched?

None. I modified the `== 'exec'` case and run the regression tests.

Unfortunately I forgot to clean before running the make files and the manifest target doesn't depend on the signer script. I now found that there's indeed a case for address 0: position independent code (In this case setting `area.addr` to `None` does the right thing).

Anyway we can still remove the if, see PR #620. At least the Pal/shim regression tests work fine with the gaps.
